sql: rework literal constant casts

RaduBerinde opened the pull request:
Literal constants have special type-checking code which allows them to be
implicitly cast to various types. The problem is that these conversions happen
very early and are invisible to the optimizer, and the result of some
conversions is not universal: it can depend on the current timezone and even the
current time (e.g. 'now' is a valid timestamp). This interacts badly with plan
caching (see #48717).

This commit changes the problematic conversions to return a `CastExpr` instead
of converting all the way to `Datum`. The conversion then happens at evaluation
time, or during constant folding in the optimizer. Soon the optimizer will have
special support for folding stable casts and operations in a way that works
correctly with caching.

We also improve the type-checking code in `CastExpr` to eliminate the cast in
all cases where the input and output types are `Identical`. This is a much more
reliable way of deciding when to elide the cast and also simplifies many
expressions that had unnecessary casts before.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50172)
<!-- Reviewable:end -->

RaduBerinde commented:
___
*[pkg/sql/logictest/testdata/logic_test/enums, line 421 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50172#-M9f7FKDDs72JlIHPSLh:-M9f7FKDDs72JlIHPSLi:b-qzljyl) ([raw file](https://github.com/cockroachdb/cockroach/blob/72d1ec4b9974c8ecd755f43c77448b59ae6706d9/pkg/sql/logictest/testdata/logic_test/enums#L421)):*
> ```
> 
> # Test that we can't mix enums in an array.
> query error pq: expected 'cockroach'::public.dbs to be of type greeting, found type dbs
> ```

CC @rohany not quite sure why there's a difference, does this look ok to you? BTW I bet we get the same message without this PR by replacing the strings with something other than a literal, e.g. `('he' || 'llo')`


<!-- Sent from Reviewable.io -->

rohany commented on pkg/sql/logictest/testdata/logic_test/enums:
```diff
@@ -418,7 +418,7 @@ SELECT ARRAY['hello']::_greeting, ARRAY['hello'::greeting]
 
 # Test that we can't mix enums in an array.
-query error pq: expected 'cockroach'::public.dbs to be of type greeting, found type dbs
+query error pq: expected 'cockroach'::dbs to be of type greeting, found type dbs
```
This is a bit worrying to me -- if the prefix has gone away, that makes me worry that there is an `TypeReference` sitting around not resolved into a `types.T` anymore.

rohany replied:
I checked out the branch, and the test is fixed by this diff -- 
```
+++ b/pkg/sql/sem/tree/type_check.go
@@ -468,6 +468,7 @@ func (expr *CastExpr) TypeCheck(
        if err != nil {
                return nil, err
        }
+       expr.Type = exprType
        switch {
        case isConstant(expr.Expr):
                c := expr.Expr.(Constant)
```
It should have been done like this in the first place. We need to be setting `expr.Type` to the resolved types.T during type checking wherever expressions contain `TypeReferences`. This is so that they can be formatted correctly in different contexts -- the unresolved name doesn't give us any power. Your patch exposed the bug by returning early if the cast types were identical.

rohany commented on pkg/sql/sem/tree/constant.go:
```diff
@@ -545,10 +550,29 @@ func (expr *StrVal) ResolveAsType(ctx *SemaContext, typ *types.T) (Datum, error)
+		}
+		expr.resString = DString(expr.s)
+		c := NewTypedCastExpr(&expr.resString, typ)
+		return c.TypeCheck(context.TODO(), ctx, typ)
```
It's important to get the context down into here -- it's for k/v lookups performed by type resolution.

RaduBerinde replied:
Ah, thanks, good catch! I meant to do that and forgot about it.

rohany commented on pkg/sql/sem/tree/constant.go:
```diff
@@ -545,10 +550,29 @@ func (expr *StrVal) ResolveAsType(ctx *SemaContext, typ *types.T) (Datum, error)
 
-	datum, err := ParseAndRequireString(typ, expr.s, ctx)
-	return datum, err
+	case types.ArrayFamily,
```
Is array included in here because it could be an array of these "bad" types? If so, can we make this check more granular to avoid doing the array cast at runtime for say int arrays?

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50172)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @otan, and @rohany)

___
*[pkg/sql/logictest/testdata/logic_test/enums, line 421 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50172#-M9f7AqHmoNLH9CaXNXl-r1-421:-M9ftbP6-S-JYySJgRMX:b1ldyj0) ([raw file](https://github.com/cockroachdb/cockroach/blob/63507a4ee545f827d624b9f6c9c5b3f430d5e315/pkg/sql/logictest/testdata/logic_test/enums#L421)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I checked out the branch, and the test is fixed by this diff -- 
```
index 25b9df84a1..133c262f19 100644
--- a/pkg/sql/sem/tree/type_check.go
+++ b/pkg/sql/sem/tree/type_check.go
@@ -468,6 +468,7 @@ func (expr *CastExpr) TypeCheck(
        if err != nil {
                return nil, err
        }
+       expr.Type = exprType
        switch {
        case isConstant(expr.Expr):
                c := expr.Expr.(Constant)
@@ -510,7 +511,6 @@ func (expr *CastExpr) TypeCheck(
        if ok, c := isCastDeepValid(castFrom, exprType); ok {
                telemetry.Inc(c)
                expr.Expr = typedSubExpr
-               expr.Type = exprType
                expr.typ = exprType
                return expr, nil
        }
```
It should have been done like this in the first place (my bad). We need to be setting `expr.Type` to the resolved types.T during type checking wherever expressions contain `TypeReferences`. This is so that they can be formatted correctly in different contexts -- the unresolved name doesn't give us any power. Your patch exposed the bug by returning early if the cast types were identical.
</blockquote></details>

Nice, thank you so much for taking a look! Done. Unfortunately it's one more case of in-place AST modification :(
___
*[pkg/sql/sem/tree/constant.go, line 554 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50172#-M9f7AqIHD_TK_Az9WQL-r1-554:-M9ftCLb7dpp1RW0onyb:bchp454) ([raw file](https://github.com/cockroachdb/cockroach/blob/63507a4ee545f827d624b9f6c9c5b3f430d5e315/pkg/sql/sem/tree/constant.go#L554)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Is array included in here because it could be an array of these "bad" types? If so, can we make this check more granular to avoid doing the array cast at runtime for say int arrays?
</blockquote></details>

We could (it would have to be something recursive), but it shouldn't matter much.. The optimizer has its own rules for folding casts and will fold anything that can be folded.
___
*[pkg/sql/sem/tree/constant.go, line 571 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50172#-M9f7AqIHD_TK_Az9WQL-r1-571:-M9fvDNTAK8oW3DozAUH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/63507a4ee545f827d624b9f6c9c5b3f430d5e315/pkg/sql/sem/tree/constant.go#L571)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Ah, thanks, good catch! I meant to do that and forgot about it.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rohany reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50172)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @otan, and @RaduBerinde)

___
*[pkg/sql/logictest/testdata/logic_test/enums, line 421 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50172#-M9f7AqHmoNLH9CaXNXl-r1-421:-M9jKQeF-93p7yyyadHG:blgqykn) ([raw file](https://github.com/cockroachdb/cockroach/blob/63507a4ee545f827d624b9f6c9c5b3f430d5e315/pkg/sql/logictest/testdata/logic_test/enums#L421)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Nice, thank you so much for taking a look! Done. Unfortunately it's one more case of in-place AST modification :(
</blockquote></details>

NP. Yeah, I'm not too happy about it, but it seems unavoidable nonetheless.
___
*[pkg/sql/sem/tree/constant.go, line 554 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50172#-M9f7AqIHD_TK_Az9WQL-r1-554:-M9jKi5X2hbqIcgahqX2:bigpj1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/63507a4ee545f827d624b9f6c9c5b3f430d5e315/pkg/sql/sem/tree/constant.go#L554)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We could (it would have to be something recursive), but it shouldn't matter much.. The optimizer has its own rules for folding casts and will fold anything that can be folded.
</blockquote></details>

Ok, sounds good.


<!-- Sent from Reviewable.io -->

otan commented on pkg/sql/sem/tree/constant.go:
```diff
@@ -545,10 +554,29 @@ func (expr *StrVal) ResolveAsType(ctx *SemaContext, typ *types.T) (Datum, error)
+		// later. So in this case we return a CastExpr and let the conversion happen
+		// at evaluation time. We still want to error out if the conversion is not
+		// possible though.
+		_, err := ParseAndRequireString(typ, expr.s, semaCtx)
```
can we re-use the cast volatility information here to be more exact?

RaduBerinde replied:
Good call, done.

RaduBerinde committed:
sql: rework literal constant casts

Literal constants have special type-checking code which allows them to be
implicitly cast to various types. The problem is that these conversions happen
very early and are invisible to the optimizer, and the result of some
conversions is not universal: it can depend on the current timezone and even the
current time (e.g. 'now' is a valid timestamp). This interacts badly with plan
caching (see #48717).

This commit changes the problematic conversions to return a `CastExpr` instead
of converting all the way to `Datum`. The conversion then happens at evaluation
time, or during constant folding in the optimizer. Soon the optimizer will have
special support for folding stable casts and operations in a way that works
correctly with caching.

We also improve the type-checking code in `CastExpr` to eliminate the cast in
all cases where the input and output types are `Identical`. This is a much more
reliable way of deciding when to elide the cast and also simplifies many
expressions that had unnecessary casts before.

Release note: None

RaduBerinde commented:
bors r+
